stdin and stdout of existing processes are lost after a restart of nomad

# Issue
Nomad is successfully able to reattach a job using `nomad-driver-containerd` after a restart but `stdin` and `stdout` are lost in the process.

# Versions
`nomad-driver-containerd` `v0.9.1`
`nomad` `v1.1.4`

# How to reproduce

  1. Consider the following changes to the Vagrant agent configuration to have Nomad run in standard mode, not `dev` mode to persist state.
```
$ cat example/agent.hcl 
log_level = "DEBUG"
data_dir = "/tmp/nomad"

plugin "containerd-driver" {
  config {
    enabled = true
    containerd_runtime = "io.containerd.runc.v2"
    stats_interval = "5s"
  }
}

server {
  enabled = true
  bootstrap_expect = 1
  default_scheduler_config {
    scheduler_algorithm = "spread"
    memory_oversubscription_enabled = true

    preemption_config {
      batch_scheduler_enabled   = true
      system_scheduler_enabled  = true
      service_scheduler_enabled = true
    }
  }
}

client {
  enabled = true
  host_volume "s1" {
    path = "/tmp/host_volume/s1"
    read_only = false
  }
}
```

2. Run the Hello job
```
$ nomad job run example/hello.hcl
```

3. See new log lines are appended every 3 seconds
```
$ nomad alloc logs -f -job hello
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
```
4. Restart nomad
```
$ sudo systemctl restart nomad
```
5. See logs aren't appended anymore despite the loop still running
```
$ nomad alloc logs -f -job hello
Hello world: sleeping for 3 seconds.
Hello world: sleeping for 3 seconds.
```

`ps faux` from the guest
```
$ nomad alloc exec -job hello /bin/ps faux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root       221  0.0  0.1  34424  2752 pts/0    Rs+  23:11   0:00 /bin/ps faux
root         1  0.0  0.1  18040  2956 ?        Ss   23:01   0:00 /bin/bash /tmp/
root       220  0.0  0.0   4376   648 ?        S    23:11   0:00 sleep 3
```
`ps faux` from the host
```
root      9112  0.0  0.4 112244  8636 ?        Sl   23:01   0:00 /usr/local/bin/containerd-shim-runc-v2 -namespace nomad -id hello-task-ea92913f-1951-8117-6f4c-b2fcd28e898b -address /run/containerd/containerd.sock
root      9138  0.0  0.1  18040  2956 ?        Ss   23:01   0:00  \_ /bin/bash /tmp/print.sh
root      9632  0.0  0.0   4376   672 ?        S    23:11   0:00      \_ sleep 3
```
Nomad still knows about the job
```
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ nomad job status hello
ID            = hello
Name          = hello
Submit Date   = 2021-08-30T23:01:06Z
Type          = service
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group   Queued  Starting  Running  Failed  Complete  Lost
hello-group  0       0         1        0       1         0

Latest Deployment
ID          = 5bfe4965
Status      = successful
Description = Deployment completed successfully

Deployed
Task Group   Desired  Placed  Healthy  Unhealthy  Progress Deadline
hello-group  1        1       1        0          2021-08-30T23:11:18Z

Allocations
ID        Node ID   Task Group   Version  Desired  Status    Created     Modified
ea92913f  0b680b8d  hello-group  2        run      running   11m58s ago  11m47s ago
```

# Observations

Looking at open file descriptors for the containerd driver it seems that the driver is not recovering the stdin and stdout named pipes for running jobs.

**Before restart**
```
root      9253  0.0  1.8 1339240 38656 ?       Sl   23:02   0:00  \_ /tmp/nomad-driver-containerd/containerd-driver
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ sudo lsof -np 9253                                                                                                                     
COMMAND    PID USER   FD      TYPE             DEVICE SIZE/OFF    NODE NAME                                                                                                                                
container 9253 root  cwd       DIR              253,0     4096       2 /                                                                                                                                   
container 9253 root  rtd       DIR              253,0     4096       2 /                                                                                                                                   
container 9253 root  txt       REG              253,0 43373088 3801138 /tmp/nomad-driver-containerd/containerd-driver                                                                                      
container 9253 root  mem       REG              253,0   101168 2097669 /lib/x86_64-linux-gnu/libresolv-2.27.so                                                                                             
container 9253 root  mem       REG              253,0    26936 2097643 /lib/x86_64-linux-gnu/libnss_dns-2.27.so                                                                                            
container 9253 root  mem       REG              253,0    47568 2097645 /lib/x86_64-linux-gnu/libnss_files-2.27.so                                                                                          
container 9253 root  mem       REG              253,0  2030544 2097578 /lib/x86_64-linux-gnu/libc-2.27.so                                                                                                  
container 9253 root  mem       REG              253,0   144976 2097665 /lib/x86_64-linux-gnu/libpthread-2.27.so                                                                                            
container 9253 root  mem       REG              253,0    14560 2097595 /lib/x86_64-linux-gnu/libdl-2.27.so                                                                                                 
container 9253 root  mem       REG              253,0   170960 2097554 /lib/x86_64-linux-gnu/ld-2.27.so                                                                                                    
container 9253 root    0r      CHR                1,3      0t0       6 /dev/null                                                                                                                           
container 9253 root    1w     FIFO               0,12      0t0   54392 pipe                                                                                                                                
container 9253 root    2w     FIFO               0,12      0t0   54393 pipe                                                                                                                                
container 9253 root    3u     unix 0xffff97c5b45a4400      0t0   54398 type=STREAM                                                                                                                         
container 9253 root    4u  a_inode               0,13        0    9576 [eventpoll]                                                                                                                         
container 9253 root    5r     FIFO               0,12      0t0   54397 pipe                                                                                                                                
container 9253 root    6w     FIFO               0,12      0t0   54397 pipe                                                                                                                                
container 9253 root    7u     unix 0xffff97c5b45a6400      0t0   54401 /tmp/plugin735931456 type=STREAM                                                                                                    
container 9253 root    8r     FIFO               0,12      0t0   54402 pipe                          
container 9253 root    9w     FIFO               0,12      0t0   54402 pipe                          
container 9253 root   10r     FIFO               0,12      0t0   54403 pipe                          
container 9253 root   11w     FIFO               0,12      0t0   54403 pipe                          
container 9253 root   12u     unix 0xffff97c5b45a7c00      0t0   54242 /tmp/plugin735931456 type=STREAM
container 9253 root   15u     FIFO              253,0      0t0 3801166 /tmp/nomad/alloc/20294144-56b6-fc55-e971-02028f7c7e2a/alloc/logs/.hello-task.stdout.fifo                                            
container 9253 root   16u     FIFO              253,0      0t0 3801168 /tmp/nomad/alloc/20294144-56b6-fc55-e971-02028f7c7e2a/alloc/logs/.hello-task.stderr.fifo                                            
container 9253 root   17u     FIFO               0,22      0t0     706 /run/containerd/fifo/739129716/hello-task-20294144-56b6-fc55-e971-02028f7c7e2a-stdout
container 9253 root   18u     FIFO               0,22      0t0     710 /run/containerd/fifo/739129716/hello-task-20294144-56b6-fc55-e971-02028f7c7e2a-stderr
container 9253 root   19r     FIFO               0,22      0t0     706 /run/containerd/fifo/739129716/hello-task-20294144-56b6-fc55-e971-02028f7c7e2a-stdout
container 9253 root   20r     FIFO               0,22      0t0     710 /run/containerd/fifo/739129716/hello-task-20294144-56b6-fc55-e971-02028f7c7e2a-stderr
```
**After restart**
```
root     10218  0.3  1.5 1259156 32552 ?       Sl   23:21   0:00  \_ /tmp/nomad-driver-containerd/containerd-driver
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ sudo lsof -np 10218
COMMAND     PID USER   FD      TYPE             DEVICE SIZE/OFF    NODE NAME
container 10218 root  cwd       DIR              253,0     4096       2 /
container 10218 root  rtd       DIR              253,0     4096       2 /
container 10218 root  txt       REG              253,0 43373088 3801138 /tmp/nomad-driver-containerd/containerd-driver
container 10218 root  mem       REG              253,0  2030544 2097578 /lib/x86_64-linux-gnu/libc-2.27.so
container 10218 root  mem       REG              253,0   144976 2097665 /lib/x86_64-linux-gnu/libpthread-2.27.so
container 10218 root  mem       REG              253,0    14560 2097595 /lib/x86_64-linux-gnu/libdl-2.27.so
container 10218 root  mem       REG              253,0   170960 2097554 /lib/x86_64-linux-gnu/ld-2.27.so
container 10218 root    0r      CHR                1,3      0t0       6 /dev/null
container 10218 root    1w     FIFO               0,12      0t0   61256 pipe
container 10218 root    2w     FIFO               0,12      0t0   61257 pipe
container 10218 root    3u     unix 0xffff97c5b16a4400      0t0   61273 type=STREAM
container 10218 root    4u  a_inode               0,13        0    9576 [eventpoll]
container 10218 root    5r     FIFO               0,12      0t0   61272 pipe
container 10218 root    6w     FIFO               0,12      0t0   61272 pipe
container 10218 root    7u     unix 0xffff97c5b16a6c00      0t0   61274 /tmp/plugin654816476 type=STREAM
container 10218 root    8r     FIFO               0,12      0t0   61275 pipe
container 10218 root    9w     FIFO               0,12      0t0   61275 pipe
container 10218 root   10r     FIFO               0,12      0t0   61276 pipe
container 10218 root   11w     FIFO               0,12      0t0   61276 pipe
container 10218 root   12u     unix 0xffff97c5b16f6800      0t0   61278 /tmp/plugin654816476 type=STREAM
```

Thanks for looking at this @shishir-a412ed 
I had forgotten a step in my reproduce to remove `-dev` from the Nomad systemd unit file in `vagrant/setup.sh`.
This change is key as in `dev` mode Nomad tears down all the jobs upon receiving `SIGTERM` so the `RecoverTask` function does not get called when Nomad comes back up (it might be a new behavior, not sure).
To make things clearer, I pushed a branch with all the required changes and I ran your steps to make sure I was still able to reproduce: https://github.com/th0m/nomad-driver-containerd/commit/73cd9fc2ffe4d365698585c01bd9e3d8e0a21b42

Here is the output after `vagrant destroy`, `vagrant up` and `vagrant ssh`:
```
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ nomad job run example/hello.nomad                                                          
==> 2021-09-02T18:15:34Z: Monitoring evaluation "975d60ce"                                                                                                     
    2021-09-02T18:15:34Z: Evaluation triggered by job "hello"                                                                                                  
==> 2021-09-02T18:15:35Z: Monitoring evaluation "975d60ce"                                                                                                     
    2021-09-02T18:15:35Z: Evaluation within deployment: "b57e2538"                                                                                             
    2021-09-02T18:15:35Z: Allocation "fcb0348a" created: node "2f7a1a27", group "hello-group"                                                                  
    2021-09-02T18:15:35Z: Evaluation status changed: "pending" -> "complete"                                                                                   
==> 2021-09-02T18:15:35Z: Evaluation "975d60ce" finished with status "complete"                                                                                
==> 2021-09-02T18:15:35Z: Monitoring deployment "b57e2538"                                                                                                     
  ✓ Deployment "b57e2538" successful                                                                                                                           
                                                                                                                                                               
    2021-09-02T18:15:58Z                                                                                                                                       
    ID          = b57e2538                                                                                                                                     
    Job ID      = hello                                                                                                                                        
    Job Version = 0                                                                                                                                            
    Status      = successful                                                                                                                                   
    Description = Deployment completed successfully                                                                                                            
                                                                               
    Deployed                                                                   
    Task Group   Desired  Placed  Healthy  Unhealthy  Progress Deadline                                                                                        
    hello-group  1        1       1        0          2021-09-02T18:25:56Z     
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ nomad job status hello                                                                     
ID            = hello                                                                                                                                          
Name          = hello                                                          
Submit Date   = 2021-09-02T18:15:34Z                                           
Type          = service                                                                                                                                        
Priority      = 50                                                             
Datacenters   = dc1                                                                                                                                            
Namespace     = default                                                                                                                                        
Status        = running                                                        
Periodic      = false                                                          
Parameterized = false                                                                                                                                          
                                                                               
Summary                                                                        
Task Group   Queued  Starting  Running  Failed  Complete  Lost                                                                                                 
hello-group  0       0         1        0       0         0           
                                                                               
Latest Deployment                                                              
ID          = b57e2538                                                         
Status      = successful                                                                                                                                       
Description = Deployment completed successfully                                                                                                                
                                                                                                                                                               
Deployed                                                                       
Task Group   Desired  Placed  Healthy  Unhealthy  Progress Deadline
hello-group  1        1       1        0          2021-09-02T18:25:56Z
                                                                               
Allocations                                                                    
ID        Node ID   Task Group   Version  Desired  Status   Created  Modified
fcb0348a  2f7a1a27  hello-group  0        run      running  35s ago  12s ago
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ nomad alloc logs fcb0348a                                                                  
Count is: 0                                                                    
Count is: 1                                                                    
Count is: 2                                                                    
Count is: 3                                                                                                                                                    
Count is: 4                                                                                                                                                    
Count is: 5                                                                                                                                                    
Count is: 6                                                                                                                                                    
Count is: 7                                                                                                                                                    
Count is: 8                                                                                                                                                    
Count is: 9                                                                    
Count is: 10                                                                                                                                                   
Count is: 11                                                                                                                                                   
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ systemctl restart nomad                                                                    
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ===           
Authentication is required to restart 'nomad.service'.                                                                                                         
Authenticating as: vagrant,,, (vagrant)                                                                                                                        
Password:                                                                      
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ sudo systemctl restart nomad
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ sudo systemctl status nomad                                                                
● nomad.service - nomad client + nomad server + nomad-driver-containerd
   Loaded: loaded (/lib/systemd/system/nomad.service; disabled; vendor preset: enabled)                                                                        
   Active: active (running) since Thu 2021-09-02 18:16:21 UTC; 8s ago
     Docs: https://nomadproject.io                                             
 Main PID: 15923 (nomad)                                                       
    Tasks: 16                                                                  
   CGroup: /system.slice/nomad.service                                         
           ├─15923 /usr/bin/nomad agent -bind=0.0.0.0 -config=/home/vagrant/go/src/github.com/Roblox/nomad-driver-containerd/example/agent.hcl -plugin-dir=/tm 
           └─15958 /tmp/nomad-driver-containerd/containerd-driver

Sep 02 18:16:27 vagrant nomad[15923]:     2021-09-02T18:16:27.476Z [DEBUG] worker.service_sched: reconciled current state with desired state: eval_id=d141eb1c 
Sep 02 18:16:27 vagrant nomad[15923]: Desired Changes for "hello-group": (place 0) (inplace 0) (destructive 0) (stop 0) (migrate 0) (ignore 1) (canary 0)"     
Sep 02 18:16:27 vagrant nomad[15923]:     2021-09-02T18:16:27.476Z [DEBUG] worker.service_sched: setting eval status: eval_id=d141eb1c-db3a-a9a8-3d78-755097c7 
Sep 02 18:16:27 vagrant nomad[15923]:     2021-09-02T18:16:27.480Z [DEBUG] worker: updated evaluation: eval="<Eval "d141eb1c-db3a-a9a8-3d78-755097c79d70" JobI 
Sep 02 18:16:27 vagrant nomad[15923]:     2021-09-02T18:16:27.480Z [DEBUG] worker: ack evaluation: eval_id=d141eb1c-db3a-a9a8-3d78-755097c79d70                
Sep 02 18:16:28 vagrant nomad[15923]:     2021-09-02T18:16:28.502Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: reattached plugin process exited: 
Sep 02 18:16:28 vagrant nomad[15923]:     2021-09-02T18:16:28.502Z [ERROR] client.alloc_runner.task_runner.task_hook: failed to start logmon: alloc_id=fcb0348 
Sep 02 18:16:28 vagrant nomad[15923]:     2021-09-02T18:16:28.502Z [WARN]  client.alloc_runner.task_runner.task_hook: logmon shutdown while making request: al 
Sep 02 18:16:28 vagrant nomad[15923]:     2021-09-02T18:16:28.502Z [WARN]  client.alloc_runner.task_runner.task_hook: logmon shutdown while making request; re 
Sep 02 18:16:28 vagrant nomad[15923]:     2021-09-02T18:16:28.502Z [DEBUG] client.alloc_runner.task_runner.task_hook.logmon: plugin exited: alloc_id=fcb0348a- 
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ nomad job status hello                                                                     
ID            = hello                                                          
Name          = hello                                                          
Submit Date   = 2021-09-02T18:15:34Z                                           
Type          = service                                                        
Priority      = 50                                                             
Datacenters   = dc1                                                            
Namespace     = default                                                        
Status        = running                                                        
Periodic      = false                                                          
Parameterized = false                                                          

Summary                                                                        
Task Group   Queued  Starting  Running  Failed  Complete  Lost
hello-group  0       0         1        0       0         0

Latest Deployment                                                              
ID          = b57e2538                                                         
Status      = successful                                                       
Description = Deployment completed successfully

Deployed                                                                       
Task Group   Desired  Placed  Healthy  Unhealthy  Progress Deadline
hello-group  1        1       1        0          2021-09-02T18:25:56Z

Allocations                                                                    
ID        Node ID   Task Group   Version  Desired  Status   Created   Modified
fcb0348a  2f7a1a27  hello-group  0        run      running  1m3s ago  40s ago
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ nomad alloc logs -f fcb0348a                                                               
Count is: 0
Count is: 1
Count is: 2
Count is: 3
Count is: 4
Count is: 5
Count is: 6
Count is: 7
Count is: 8
Count is: 9
Count is: 10
Count is: 11
Count is: 12
Count is: 13
Count is: 14
vagrant@vagrant:~/go/src/github.com/Roblox/nomad-driver-containerd$ nomad status                              
ID     Type     Priority  Status   Submit Date                                                                                                                 
hello  service  50        running  2021-09-02T18:15:34Z
root@vagrant:/home/vagrant/go/src/github.com/Roblox/nomad-driver-containerd# export CONTAINERD_NAMESPACE=nomad
root@vagrant:/home/vagrant/go/src/github.com/Roblox/nomad-driver-containerd# ./nerdctl ps
CONTAINER ID    IMAGE                        COMMAND            CREATED          STATUS    PORTS    NAMES
hello-task-f    docker.io/shm32/count:1.0    "/tmp/count.sh"    7 minutes ago    Up
```

@th0m Aah you are right! (I am able to reproduce the issue now)

```
This change is key as in dev mode Nomad tears down all the jobs upon receiving SIGTERM so the RecoverTask function does not get called when Nomad comes back up (it might be a new behavior, not sure).
```
Yeah, seems like a regression. I use to test this code path (Recover task in `containerd-driver`) in Nomad `dev` mode in earlier versions, and it used to work nicely. Seems like something changed in the later versions which broke this! Will raise an issue with `hashicorp/nomad`.

For this bug, I see this in logs (Will keep digging why it's happening) after `systemctl restart nomad`

```
Sep 02 18:46:24 vagrant nomad[16214]:     2021-09-02T18:46:24.593Z [ERROR] client.alloc_runner.task_runner.task_hook: failed to start logmon: alloc_id=ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d task=hello-task error="plugin is shut down"
Sep 02 18:46:24 vagrant nomad[16214]:     2021-09-02T18:46:24.593Z [WARN]  client.alloc_runner.task_runner.task_hook: logmon shutdown while making request: alloc_id=ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d task=hello-task error="plugin is shut down"
Sep 02 18:46:24 vagrant nomad[16214]:     2021-09-02T18:46:24.593Z [WARN]  client.alloc_runner.task_runner.task_hook: logmon shutdown while making request; retrying: alloc_id=ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d task=hello-task attempts=1 error="plugin is shut down"
Sep 02 18:46:25 vagrant nomad[16214]:     2021-09-02T18:46:25.605Z [INFO]  client.alloc_runner.task_runner.task_hook.logmon.nomad: opening fifo: alloc_id=ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d task=hello-task path=/tmp/nomad/alloc/ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d/alloc/logs/.hello-task.stdout.fifo @module=logmon timestamp=2021-09-02T18:46:25.604Z
Sep 02 18:46:25 vagrant nomad[16214]:     2021-09-02T18:46:25.605Z [INFO]  client.alloc_runner.task_runner.task_hook.logmon.nomad: opening fifo: alloc_id=ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d task=hello-task path=/tmp/nomad/alloc/ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d/alloc/logs/.hello-task.stderr.fifo @module=logmon timestamp=2021-09-02T18:46:25.605Z
```

After I restart `nomad`, the stdout/stderr fifos are still there (as expected)

```
root@vagrant:/tmp/nomad/alloc/ca5e1680-4ac1-6c3b-0d29-f3b1dc145d5d/alloc/logs# ls
hello-task.stderr.0  hello-task.stdout.0
```
However, either `Nomad` or `containerd-driver` have trouble reattaching to them. Need to look into why this is happening!

--- a/containerd/containerd.go
+++ b/containerd/containerd.go
@@ -341,12 +341,7 @@ func (d *Driver) loadContainer(id string) (containerd.Container, error) {
 }
 
 func (d *Driver) createTask(container containerd.Container, stdoutPath, stderrPath string) (containerd.Task, error) {
-	stdout, err := openFIFO(stdoutPath)
-	if err != nil {
-		return nil, err
-	}
-
-	stderr, err := openFIFO(stderrPath)
+	stdout, stderr, err := getStdoutStderrFifos(stdoutPath, stderrPath)
 	if err != nil {
 		return nil, err
 	}
@@ -357,9 +352,14 @@ func (d *Driver) createTask(container containerd.Container, stdoutPath, stderrPa
 	return container.NewTask(ctxWithTimeout, cio.NewCreator(cio.WithStreams(nil, stdout, stderr)))
 }
 
-func (d *Driver) getTask(container containerd.Container) (containerd.Task, error) {
+func (d *Driver) getTask(container containerd.Container, stdoutPath, stderrPath string) (containerd.Task, error) {
+	stdout, stderr, err := getStdoutStderrFifos(stdoutPath, stderrPath)
+	if err != nil {
+		return nil, err
+	}
+
 	ctxWithTimeout, cancel := context.WithTimeout(d.ctxContainerd, 30*time.Second)
 	defer cancel()
 
-	return container.Task(ctxWithTimeout, cio.Load)
+	return container.Task(ctxWithTimeout, cio.NewAttach(cio.WithStreams(nil, stdout, stderr)))
 }

--- a/containerd/driver.go
+++ b/containerd/driver.go
@@ -207,6 +207,8 @@ type TaskConfig struct {
 type TaskState struct {
 	StartedAt     time.Time
 	ContainerName string
+	StdoutPath    string
+	StderrPath    string
 }
 
 type Driver struct {
@@ -496,6 +498,8 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
 	driverState := TaskState{
 		StartedAt:     h.startedAt,
 		ContainerName: containerName,
+		StdoutPath:    cfg.StdoutPath,
+		StderrPath:    cfg.StderrPath,
 	}
 
 	if err := handle.SetDriverState(&driverState); err != nil {
@@ -539,7 +543,7 @@ func (d *Driver) RecoverTask(handle *drivers.TaskHandle) error {
 		return fmt.Errorf("Error in recovering container: %v", err)
 	}
 
-	task, err := d.getTask(container)
+	task, err := d.getTask(container, taskState.StdoutPath, taskState.StderrPath)
 	if err != nil {
 		return fmt.Errorf("Error in recovering task: %v", err)
 	}

--- a/containerd/utils.go
+++ b/containerd/utils.go
@@ -37,6 +37,20 @@ func buildMountpoint(mountType, mountTarget, mountSource string, mountOptions []
 	return m
 }
 
+// getStdoutStderrFifos return the container's stdout and stderr FIFO's.
+func getStdoutStderrFifos(stdoutPath, stderrPath string) (*os.File, *os.File, error) {
+	stdout, err := openFIFO(stdoutPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	stderr, err := openFIFO(stderrPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	return stdout, stderr, nil
+}
+
 // FIFO's are named pipes in linux.
 // openFIFO() opens the nomad task stdout/stderr pipes and returns the fd.
 func openFIFO(path string) (*os.File, error) {

--- a/tests/run_tests.sh
+++ b/tests/run_tests.sh
@@ -29,6 +29,7 @@ main() {
 	setup
 	echo "INFO: Checking if nomad-driver-containerd is up and running, and nomad is ready to accept jobs."
 	is_containerd_driver_active
+	is_nomad_ready
 
 	run_tests $@
 	exit $PASS_STATUS
@@ -200,6 +201,29 @@ is_containerd_driver_active() {
 		echo "ERROR: containerd driver didn't come up. exit 1."
 		exit 1
 	fi
+}
+
+is_nomad_ready() {
+        i="0"
+        while test $i -lt 5
+        do
+                set +e
+                status=$(curl -s http://127.0.0.1:4646/v1/nodes|jq '.[0] ."Status"')
+                rc=$?
+                set -e
+                if [[ $rc -eq 0 && $status = \"ready\" ]]; then
+                        echo "INFO: nomad is ready to accept jobs."
+                        break
+                fi
+                echo "INFO: nomad is initializing, sleep for 4 seconds."
+                sleep 4s
+                i=$[$i+1]
+        done
+
+        if [ $i -ge 5 ]; then
+                echo "ERROR: nomad didn't come up. exit 1."
+                exit 1
+        fi
 }
 
 main "$@"

